src/dragonball/src/dbs_pci/src/msi.rs

```diff
@@ -0,0 +1,759 @@
+// Copyright (C) 2023 Alibaba Cloud. All rights reserved.
+// Copyright © 2019 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+use std::sync::Arc;
+
+use byteorder::{ByteOrder, LittleEndian};
+use dbs_interrupt::{
+    DeviceInterruptManager, DeviceInterruptMode, InterruptIndex, InterruptManager,
+    InterruptSourceGroup,
+};
+use log::{debug, error};
+
+use crate::configuration::{PciCapability, PciCapabilityID};
+use crate::fill_config_data;
+
+// MSI control masks
+pub(crate) const MSI_CTL_ENABLE: u16 = 0x1;
+const MSI_CTL_MULTI_MSG_ENABLE: u16 = 0x70;
+pub(crate) const MSI_CTL_64_BITS: u16 = 0x80;
+const MSI_CTL_PER_VECTOR: u16 = 0x100;
+
+// MSI message offsets
+const MSI_MSG_CTL_OFFSET: u64 = 0x2;
+const MSI_MSG_ADDR_LO_OFFSET: u64 = 0x4;
+
+// MSI message masks
+const MSI_MSG_ADDR_LO_MASK: u32 = 0xffff_fffc;
+
+/// Get number of vectors enabled from PCI MSI control register.
+pub fn msi_num_enabled_vectors(msg_ctl: u16) -> u16 {
+    let field = (msg_ctl >> 4) & 0x7;
+
+    if field > 5 {
+        return 0;
+    }
+
+    1 << field
```

Comment:
const added for 4 and 0x7;
comment added for the meaning of 5.
"since we mask down to 3 bit for number of msi enabled, and 3 bit should not be larger than 5, if that happens, meaning that there s some unknown behavior so we return 0."